### ISSUE TITLE
Restart without "close -> open" loop

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#29856]
### Problem

No way to restart without fully closing and starting with the same command.

See also vim/vim#13904. Decided to open a separate issue here as it seems that implementation will be different and the Vim issue got stale.

### Expected behavior

Available either a `:restart` command or `vim.api.nvim_restart()` function which can restart an instance without "close-reopen" loop.

The effect is expected to be at least the same as (in layman language):
- Do Neovim closing process as normal *but* stop just before actually terminating the process.
- Execute startup sequence as if a new process has started. This can also result in preserving "startup related" data like argument list, CLI options, etc. Maybe even tabpages/windows/buffers for better interactive experience.

I see this feature as useful in at least for following use cases:
- Clean reloading of user config after there were some changes made.
- Restarting instance through RPC without dealing with "close and drop connection -> reopen with proper arguments -> reconnect".

**Comment by lewis6991** [comment]
Looks like a duplicate of https://github.com/neovim/neovim/issues/19681

**Comment by echasnovski** [comment]
Indeed, missed that.

However:
- Since that issue was closed, there were rewrites of how Neovim startup process works. I specifically thought that after Neovim has TUI run in a separate process, this might be not that complicated to implement. Cost-benefit analysis might have shifted here. Hence the "needs:discussion" label.
- The benefit of available restart through RPC seems to be a valid argument for having it in core.
- Suggestion to move this functionality to plugin (which are mostly meant to work *inside* current session) to me seems a bit far-fetched.

**Comment by justinmk** [comment]
Closing in favor of https://github.com/neovim/neovim/issues/32484 , which is a more narrowly-defined spec.

If something more elaborate than that is wanted, we can reopen this one and track that here.
